### DISCUSSION TITLE
Elastic Agent Error with FQDN

### DISCUSSION THREAD

**Discussion by EddieN17** [Security-Onion-Solutions/securityonion#10991] [2.4] [answered]
Hello,

This seems to be a similar issue of https://github.com/Security-Onion-Solutions/securityonion/discussions/10504, but that was supposedly solved with one of the updates, but I'm getting this error again. The Elastic Agent on the manager or any other attempts of an Elastic Agent elsewhere can't connect to the fleet server. 

Running the SO 2.4.5 ISO, has access to the internet but through a proxy. When prompted on how to access the web interface, I selected "other" and entered my FQDN. The FQDN is set up correctly within my DNS resolver. 

SO Setup Error Log
```
[ERROR   ] stderr:
Error: enroll command failed with exit code: 1
[ERROR   ] retcode: 1
[ERROR   ] {'pid': 143216, 'retcode': 1, 'stdout': '\nInstallation initiated, view install log for further details.', 'stderr': '\n\nInstalling in non-interactive mode.{"log.level":"info","@timestamp":"2023-08-09T00:38:43.793Z","log.origin":{"file.name":"cmd/enroll_cmd.go","file.line":478},"message":"Starting enrollment to URL: https://managerdev.*domainname*:8220/","ecs.version":"1.6.0"}\nError: fail to enroll: fail to execute request to fleet-server: Forbidden\nFor help, please see our troubleshooting guide at https://www.elastic.co/guide/en/fleet/8.8/fleet-troubleshooting.html\nError: enroll command failed with exit code: 1\nFor help, please see our troubleshooting guide at https://www.elastic.co/guide/en/fleet/8.8/fleet-troubleshooting.html\n: exit status 1'}
      Result: False
                  Error: enroll command failed with exit code: 1
```

When trying to install a Windows agent:
```
Error: fail to enroll: fail to execute request to fleet-server: x509: certificate is valid for managerdev, not managerdev.*domainname*
```

When I test the self-signed certs generated by Security Onion, the common name of the certs are assigned to just the hostname, not the FQDN.

**Comment by defensivedepth** [comment]
Is everything else working as expected?

Please run the following on the Manager command line and post a sanitized output:

`sudo salt-call pillar.get global`

**Reply by EddieN17** [reply]
Yep, everything else is working. The only issue is any Security Onion grid nodes won't connect via elastic agent to the Fleet Server and attempts with different devices using the SO Elastic Agent download fail as well, even after adding the IPs to the firewall. Only agent that shows up in the Elastic Fleet page on Kibana is the Fleet Server.
```
local:
    ----------
    airgap:
        False
    endgamehost:
    fleet_grid_enrollment_token_general:
        ekNCejJJxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    fleet_grid_enrollment_token_heavy:
        emlCejJJxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    ids:
        Suricata
    imagerepo:
        security-onion-solutions
    influxdb_host:
        managerdev
    managerip:
        172.xx.xx.xxx
    mdengine:
        ZEEK
    pipeline:
        redis
    registry_host:
        managerdev
    repo_host:
        managerdev
    soversion:
        2.4.5
    url_base:
        managerdev.xxxxxxx.xxxx.xxx
```

**Reply by EddieN17** [reply]
Also to answer the same questions that https://github.com/Security-Onion-Solutions/securityonion/discussions/10992 has.
Distributed Install Type.
ISO.
Not airgap.
Trying to run Security Onion as essentially a production environment. Multiple employees will be accessing in and out of the SOC, and having the SOC Web URL be the FQDN instead of the IP address would just make things look better. Essentially not the biggest problem in my world if I have to make the WEB URL the IP address (because if I do that on install everything runs properly), but it would be preferred to have the FQDN.

**Comment by jaypfe** [comment]
I found this issue as well. 

The workaround I have used is just going with IP in the setup, then changing the web access URL by going Configuration –> global –> url_base. This seems to work good for me. 

https://docs.securityonion.net/en/2.4/url-base.html?highlight=web%20access

**Reply by EddieN17** [reply]
I did find that workaround too, but I had a problem, with RC2 at least, since there’s now a so-elastic-fleet-urls-update that occurs (unsure when but it happened after a reboot) where it’ll change the fleet server access URLs to the url_base and any future elastic agent integrations will fail because of that as it reaches the same errors with certificates.

**Reply by jaypfe** [reply]
Ah yes I see that now. The installer will change to FQDN and fail to enroll any new agents. This did work for me in RC1

**Comment by defensivedepth** [comment, marked as answer]
@EddieN17 

I think I have figured it out. Appears that Elastic Fleet / Agent is not respecting the Cert CN when verifying the connection. Moving it to the SAN appears to have fixed it. Still testing, but if you would like to try it out, make the changes in this PR on your Manager:

https://github.com/Security-Onion-Solutions/securityonion/pull/11002

(That file is at `/opt/so/saltstack/default/salt/ssl/init.sls`). Then run the following:

`sudo salt-call state.highstate`

This should regenerate the certs & restart the containers.

**Reply by EddieN17** [reply]
I still got one issue, but that solved most of the problems. I haven't done a lot of testing yet but almost everything seems to be working properly. But is there a way to install the missing Elastic Agent on the manager? It has the FleetServer agent, but not the so-grid-nodes agent, as that failed on install.

I tried using the installer in `/opt/so/saltstack/local/salt/elasticfleet/files/so_agent-installers/` but that failed with 
```
timestamp=2023-08-09T22:36:17.125239038Z level=error message="Installation Progress" Context="Installing in non-interactive mode.{\"log.level\":\"info\",\"@timestamp\":\"2023-08-09T22:36:10.174Z\",\"log.origin\":{\"file.name\":\"cmd/enroll_cmd.go\",\"file.line\":478},\"message\":\"Starting enrollment to URL: https://managerdev.xxxxxxx.xxxx.xxx:8220/\",\"ecs.version\":\"1.6.0\"}\nError: fail to enroll: fail to execute request to fleet-server: Forbidden\nFor help, please see our troubleshooting guide at https://www.elastic.co/guide/en/fleet/8.8/fleet-troubleshooting.html\nError: enroll command failed with exit code: 1\nFor help, please see our troubleshooting guide at https://www.elastic.co/guide/en/fleet/8.8/fleet-troubleshooting.html\n" ErrorDetails="exit status 1"
```

Also important to note, I created a new manager, but before going through the install I changed the salt/ssl/init.sls file to your changes, and I still got the error on the manager, but everything else seems to be running fine. That error being:
```
[ERROR   ] {'pid': 143790, 'retcode': 1, 'stdout': '\nInstallation initiated, view install log for further details.', 'stderr': '\n\nInstalling in non-interactive mode.{"log.level":"info","@timestamp":"2023-08-09T23:38:24.730Z","log.origin":{"file.name":"cmd/enroll_cmd.go","file.line":478},"message":"Starting enrollment to URL: https://managerdev.xxxxxxx.xxxx.xxx:8220/","ecs.version":"1.6.0"}\nError: fail to enroll: fail to execute request to fleet-server: Forbidden\nFor help, please see our troubleshooting guide at https://www.elastic.co/guide/en/fleet/8.8/fleet-troubleshooting.html\nError: enroll command failed with exit code: 1\nFor help, please see our troubleshooting guide at https://www.elastic.co/guide/en/fleet/8.8/fleet-troubleshooting.html\n: exit status 1'}
      Result: False
                  Error: enroll command failed with exit code: 1
```

EDIT: To clarify as well, any elastic agent installs on endpoints or other security onion grid nodes work perfectly fine. It’s just the manager install failing for some reason.

**Reply by EddieN17** [reply]
I found the problem, but I'm unsure how to fix it. Because I set the manager up with a proxy in order to get internet access, it tries using that proxy when it installs Elastic Agent when it goes out to `https://managerdev.xxxxxxx.xxxx.xxx:8220`, but that will fail. I need the Elastic Agent to install without using my proxy configurations, given it's just on the local network and can talk to itself.

I removed the https_proxy environment variable and download elastic agent on the manager and it successfully installed, then I put the environment variable back, but that's a very duct tape solution.

Edit: New Discussion https://github.com/Security-Onion-Solutions/securityonion/discussions/11020
